File: shared/api/base_api_client.py
"""
Base API Client - Python implementation
Provides standardized HTTP communication across Python services
"""

import asyncio
import json
import logging
from abc import ABC, abstractmethod
from datetime import datetime
from typing import Any, Dict, Optional, Type, TypeVar, Union
from urllib.parse import urljoin, urlencode

import aiohttp
from aiohttp import ClientSession, ClientTimeout, ClientError

from ..errors.app_error import AppError
from ..logging.logger_service import LoggerService, LogLevel

T = TypeVar('T')


class BaseApiClient(ABC):
    """Base API Client for standardized HTTP communication"""
    
    def __init__(
        self,
        base_url: str,
        default_headers: Dict[str, str],
        timeout: int = 30,
        session: Optional[ClientSession] = None
    ):
        self.base_url = base_url.rstrip('/')
        self.default_headers = default_headers
        self.timeout = ClientTimeout(total=timeout)
        self._session = session
        self._logger = LoggerService()

    async def __aenter__(self):
        if self._session is None:
            self._session = ClientSession(
                timeout=self.timeout,
                headers=self.default_headers
            )
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self._session:
            await self._session.close()

    @property
    def session(self) -> ClientSession:
        """Get or create session"""
        if self._session is None:
            self._session = ClientSession(
                timeout=self.timeout,
                headers=self.default_headers
            )
        return self._session

    async def get(
        self,
        endpoint: str,
        headers: Optional[Dict[str, str]] = None,
        params: Optional[Dict[str, Any]] = None,
        response_type: Optional[Type[T]] = None
    ) -> Union[T, Dict[str, Any]]:
        """GET request with standardized error handling"""
        url = self._build_url(endpoint)
        request_headers = self._build_headers(headers)
        
        self._logger.log(
            level=LogLevel.INFO,
            message=f"GET {url}",
            metadata={'params': params},
            service='api_client'
        )
        
        try:
            async with self.session.get(
                url,
                headers=request_headers,
                params=params
            ) as response:
                response_data = await self._handle_response(response)
                
                if response_type:
                    return response_type.from_dict(response_data)
                return response_data
                
        except ClientError as e:
            raise AppError(
                code='NETWORK_001',
                message='Network connection failed',
                details={'original_error': str(e)}
            )
        except Exception as e:
            raise AppError(
                code='UNKNOWN_001',
                message='Unknown error occurred',
                details={'original_error': str(e)}
            )

    async def post(
        self,
        endpoint: str,
        body: Optional[Any] = None,
        headers: Optional[Dict[str, str]] = None,
        params: Optional[Dict[str, Any]] = None,
        response_type: Optional[Type[T]] = None
    ) -> Union[T, Dict[str, Any]]:
        """POST request with standardized error handling"""
        url = self._build_url(endpoint)
        request_headers = self._build_headers(headers)
        json_body = body if body is not None else {}
        
        self._logger.log(
            level=LogLevel.INFO,
            message=f"POST {url}",
            metadata={'body': body, 'params': params},
            service='api_client'
        )
        
        try:
            async with self.session.post(
                url,
                headers=request_headers,
                json=json_body,
                params=params
            ) as response:
                response_data = await self._handle_response(response)
                
                if response_type:
                    return response_type.from_dict(response_data)
                return response_data
                
        except ClientError as e:
            raise AppError(
                code='NETWORK_001',
                message='Network connection failed',
                details={'original_error': str(e)}
            )
        except Exception as e:
            raise AppError(
                code='UNKNOWN_001',
                message='Unknown error occurred',
                details={'original_error': str(e)}
            )

    async def put(
        self,
        endpoint: str,
        body: Optional[Any] = None,
        headers: Optional[Dict[str, str]] = None,
        params: Optional[Dict[str, Any]] = None,
        response_type: Optional[Type[T]] = None
    ) -> Union[T, Dict[str, Any]]:
        """PUT request with standardized error handling"""
        url = self._build_url(endpoint)
        request_headers = self._build_headers(headers)
        json_body = body if body is not None else {}
        
        self._logger.log(
            level=LogLevel.INFO,
            message=f"PUT {url}",
            metadata={'body': body, 'params': params},
            service='api_client'
        )
        
        try:
            async with self.session.put(
                url,
                headers=request_headers,
                json=json_body,
                params=params
            ) as response:
                response_data = await self._handle_response(response)
                
                if response_type:
                    return response_type.from_dict(response_data)
                return response_data
                
        except ClientError as e:
            raise AppError(
                code='NETWORK_001',
                message='Network connection failed',
                details={'original_error': str(e)}
            )
        except Exception as e:
            raise AppError(
                code='UNKNOWN_001',
                message='Unknown error occurred',
                details={'original_error': str(e)}
            )

    async def delete(
        self,
        endpoint: str,
        headers: Optional[Dict[str, str]] = None,
        params: Optional[Dict[str, Any]] = None,
        response_type: Optional[Type[T]] = None
    ) -> Union[T, Dict[str, Any]]:
        """DELETE request with standardized error handling"""
        url = self._build_url(endpoint)
        request_headers = self._build_headers(headers)
        
        self._logger.log(
            level=LogLevel.INFO,
            message=f"DELETE {url}",
            metadata={'params': params},
            service='api_client'
        )
        
        try:
            async with self.session.delete(
                url,
                headers=request_headers,
                params=params
            ) as response:
                response_data = await self._handle_response(response)
                
                if response_type:
                    return response_type.from_dict(response_data)
                return response_data
                
        except ClientError as e:
            raise AppError(
                code='NETWORK_001',
                message='Network connection failed',
                details={'original_error': str(e)}
            )
        except Exception as e:
            raise AppError(
                code='UNKNOWN_001',
                message='Unknown error occurred',
                details={'original_error': str(e)}
            )

    def _build_url(self, endpoint: str) -> str:
        """Build full URL from base URL and endpoint"""
        if endpoint.startswith('/'):
            endpoint = endpoint[1:]
        return f"{self.base_url}/{endpoint}"

    def _build_headers(self, custom_headers: Optional[Dict[str, str]] = None) -> Dict[str, str]:
        """Build headers by merging default and custom headers"""
        headers = self.default_headers.copy()
        
        if custom_headers:
            headers.update(custom_headers)
        
        # Ensure content-type is set
        if 'Content-Type' not in headers:
            headers['Content-Type'] = 'application/json'
        
        return headers

    async def _handle_response(self, response: aiohttp.ClientResponse) -> Dict[str, Any]:
        """Handle HTTP response and convert errors to AppError"""
        response_text = await response.text()
        
        self._logger.log(
            level=LogLevel.INFO,
            message=f"Response {response.status}",
            metadata={'response_body': response_text[:1000]},  # Truncate for logging
            service='api_client'
        )
        
        if 200 <= response.status < 300:
            try:
                return json.loads(response_text) if response_text else {}
            except json.JSONDecodeError as e:
                raise AppError(
                    code='PARSE_001',
                    message='Failed to parse response',
                    details={'original_error': str(e), 'response_body': response_text}
                )
        else:
            error_code = f'HTTP_{response.status}'
            error_message = f'HTTP Error {response.status}'
            error_details = {'response_body': response_text}
            
            try:
                error_data = json.loads(response_text)
                if isinstance(error_data, dict) and 'error' in error_data:
                    error_info = error_data['error']
                    error_code = error_info.get('code', error_code)
                    error_message = error_info.get('message', error_message)
                    error_details = error_info.get('details', error_details)
            except json.JSONDecodeError:
                pass  # Use default error info
            
            raise AppError(
                code=error_code,
                message=error_message,
                details=error_details
            )

    async def close(self):
        """Close the session"""
        if self._session:
            await self._session.close()


class ExtendedApiClient(BaseApiClient):
    """Extended API Client specifically for Extended Exchange"""
    
    def __init__(
        self,
        base_url: str,
        api_key: str,
        timeout: int = 30,
        session: Optional[ClientSession] = None
    ):
        super().__init__(
            base_url=base_url,
            default_headers={
                'Content-Type': 'application/json',
                'X-API-Key': api_key,
            },
            timeout=timeout,
            session=session
        )
        self.api_key = api_key

    async def get_market_data(self, symbol: str) -> Dict[str, Any]:
        """Get market data for a symbol"""
        return await self.get(
            '/market/data',
            params={'symbol': symbol}
        )

    async def place_order(self, order_data: Dict[str, Any]) -> Dict[str, Any]:
        """Place a new order"""
        return await self.post('/orders', body=order_data)

    async def get_order_history(self) -> List[Dict[str, Any]]:
        """Get order history"""
        response = await self.get('/orders/history')
        return response.get('orders', [])

    async def get_balance(self) -> Dict[str, Any]:
        """Get account balance"""
        return await self.get('/account/balance')

    async def cancel_order(self, order_id: str) -> Dict[str, Any]:
        """Cancel an order"""
        return await self.delete(f'/orders/{order_id}')

    async def get_order_status(self, order_id: str) -> Dict[str, Any]:
        """Get order status"""
        return await self.get(f'/orders/{order_id}')

File: shared/api/base_api_client.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import '../errors/app_error.dart';
import '../logging/logger_service.dart';

/// Base API Client for standardized HTTP communication
/// Provides common functionality for all API clients
abstract class BaseApiClient {
  final String baseUrl;
  final Map<String, String> defaultHeaders;
  final Duration timeout;
  final LoggerService _logger = LoggerService();

  BaseApiClient({
    required this.baseUrl,
    required this.defaultHeaders,
    this.timeout = const Duration(seconds: 30),
  });

  /// GET request with standardized error handling
  Future<T> get<T>(
    String endpoint, {
    Map<String, String>? headers,
    Map<String, String>? queryParameters,
    T Function(Map<String, dynamic>)? fromJson,
  }) async {
    final uri = _buildUri(endpoint, queryParameters);
    final requestHeaders = _buildHeaders(headers);

    _logger.logRequest('GET', uri.toString(), null);

    try {
      final response = await http
          .get(uri, headers: requestHeaders)
          .timeout(timeout);

      _logger.logResponse(response.statusCode, response.body);

      if (response.statusCode >= 200 && response.statusCode < 300) {
        final responseData = jsonDecode(response.body);
        return fromJson != null ? fromJson(responseData) : responseData as T;
      } else {
        throw _handleHttpError(response);
      }
    } on SocketException catch (e) {
      throw AppError(
        code: 'NETWORK_001',
        message: 'Network connection failed',
        details: {'original_error': e.toString()},
      );
    } on HttpException catch (e) {
      throw AppError(
        code: 'HTTP_001',
        message: 'HTTP request failed',
        details: {'original_error': e.toString()},
      );
    } on FormatException catch (e) {
      throw AppError(
        code: 'PARSE_001',
        message: 'Failed to parse response',
        details: {'original_error': e.toString()},
      );
    } catch (e) {
      throw AppError(
        code: 'UNKNOWN_001',
        message: 'Unknown error occurred',
        details: {'original_error': e.toString()},
      );
    }
  }

  /// POST request with standardized error handling
  Future<T> post<T>(
    String endpoint, {
    dynamic body,
    Map<String, String>? headers,
    Map<String, String>? queryParameters,
    T Function(Map<String, dynamic>)? fromJson,
  }) async {
    final uri = _buildUri(endpoint, queryParameters);
    final requestHeaders = _buildHeaders(headers);
    final requestBody = body != null ? jsonEncode(body) : null;

    _logger.logRequest('POST', uri.toString(), body);

    try {
      final response = await http
          .post(uri, headers: requestHeaders, body: requestBody)
          .timeout(timeout);

      _logger.logResponse(response.statusCode, response.body);

      if (response.statusCode >= 200 && response.statusCode < 300) {
        final responseData = jsonDecode(response.body);
        return fromJson != null ? fromJson(responseData) : responseData as T;
      } else {
        throw _handleHttpError(response);
      }
    } on SocketException catch (e) {
      throw AppError(
        code: 'NETWORK_001',
        message: 'Network connection failed',
        details: {'original_error': e.toString()},
      );
    } on HttpException catch (e) {
      throw AppError(
        code: 'HTTP_001',
        message: 'HTTP request failed',
        details: {'original_error': e.toString()},
      );
    } on FormatException catch (e) {
      throw AppError(
        code: 'PARSE_001',
        message: 'Failed to parse response',
        details: {'original_error': e.toString()},
      );
    } catch (e) {
      throw AppError(
        code: 'UNKNOWN_001',
        message: 'Unknown error occurred',
        details: {'original_error': e.toString()},
      );
    }
  }

  /// PUT request with standardized error handling
  Future<T> put<T>(
    String endpoint, {
    dynamic body,
    Map<String, String>? headers,
    Map<String, String>? queryParameters,
    T Function(Map<String, dynamic>)? fromJson,
  }) async {
    final uri = _buildUri(endpoint, queryParameters);
    final requestHeaders = _buildHeaders(headers);
    final requestBody = body != null ? jsonEncode(body) : null;

    _logger.logRequest('PUT', uri.toString(), body);

    try {
      final response = await http
          .put(uri, headers: requestHeaders, body: requestBody)
          .timeout(timeout);

      _logger.logResponse(response.statusCode, response.body);

      if (response.statusCode >= 200 && response.statusCode < 300) {
        final responseData = jsonDecode(response.body);
        return fromJson != null ? fromJson(responseData) : responseData as T;
      } else {
        throw _handleHttpError(response);
      }
    } on SocketException catch (e) {
      throw AppError(
        code: 'NETWORK_001',
        message: 'Network connection failed',
        details: {'original_error': e.toString()},
      );
    } on HttpException catch (e) {
      throw AppError(
        code: 'HTTP_001',
        message: 'HTTP request failed',
        details: {'original_error': e.toString()},
      );
    } on FormatException catch (e) {
      throw AppError(
        code: 'PARSE_001',
        message: 'Failed to parse response',
        details: {'original_error': e.toString()},
      );
    } catch (e) {
      throw AppError(
        code: 'UNKNOWN_001',
        message: 'Unknown error occurred',
        details: {'original_error': e.toString()},
      );
    }
  }

  /// DELETE request with standardized error handling
  Future<T> delete<T>(
    String endpoint, {
    Map<String, String>? headers,
    Map<String, String>? queryParameters,
    T Function(Map<String, dynamic>)? fromJson,
  }) async {
    final uri = _buildUri(endpoint, queryParameters);
    final requestHeaders = _buildHeaders(headers);

    _logger.logRequest('DELETE', uri.toString(), null);

    try {
      final response = await http
          .delete(uri, headers: requestHeaders)
          .timeout(timeout);

      _logger.logResponse(response.statusCode, response.body);

      if (response.statusCode >= 200 && response.statusCode < 300) {
        final responseData = jsonDecode(response.body);
        return fromJson != null ? fromJson(responseData) : responseData as T;
      } else {
        throw _handleHttpError(response);
      }
    } on SocketException catch (e) {
      throw AppError(
        code: 'NETWORK_001',
        message: 'Network connection failed',
        details: {'original_error': e.toString()},
      );
    } on HttpException catch (e) {
      throw AppError(
        code: 'HTTP_001',
        message: 'HTTP request failed',
        details: {'original_error': e.toString()},
      );
    } on FormatException catch (e) {
      throw AppError(
        code: 'PARSE_001',
        message: 'Failed to parse response',
        details: {'original_error': e.toString()},
      );
    } catch (e) {
      throw AppError(
        code: 'UNKNOWN_001',
        message: 'Unknown error occurred',
        details: {'original_error': e.toString()},
      );
    }
  }

  /// Build URI with query parameters
  Uri _buildUri(String endpoint, Map<String, String>? queryParameters) {
    final url = baseUrl.endsWith('/') ? baseUrl : '$baseUrl/';
    final fullUrl = '$url${endpoint.startsWith('/') ? endpoint.substring(1) : endpoint}';
    
    if (queryParameters != null && queryParameters.isNotEmpty) {
      return Uri.parse(fullUrl).replace(queryParameters: queryParameters);
    }
    
    return Uri.parse(fullUrl);
  }

  /// Build headers by merging default and custom headers
  Map<String, String> _buildHeaders(Map<String, String>? customHeaders) {
    final headers = Map<String, String>.from(defaultHeaders);
    
    if (customHeaders != null) {
      headers.addAll(customHeaders);
    }
    
    // Ensure content-type is set for POST/PUT requests
    if (!headers.containsKey('Content-Type')) {
      headers['Content-Type'] = 'application/json';
    }
    
    return headers;
  }

  /// Handle HTTP errors and convert to AppError
  AppError _handleHttpError(http.Response response) {
    String errorCode;
    String errorMessage;
    Map<String, dynamic>? errorDetails;

    try {
      final responseData = jsonDecode(response.body);
      
      if (responseData is Map<String, dynamic> && responseData.containsKey('error')) {
        final error = responseData['error'];
        errorCode = error['code'] ?? 'HTTP_${response.statusCode}';
        errorMessage = error['message'] ?? 'HTTP Error ${response.statusCode}';
        errorDetails = error['details'];
      } else {
        errorCode = 'HTTP_${response.statusCode}';
        errorMessage = 'HTTP Error ${response.statusCode}';
        errorDetails = {'response_body': response.body};
      }
    } catch (e) {
      errorCode = 'HTTP_${response.statusCode}';
      errorMessage = 'HTTP Error ${response.statusCode}';
      errorDetails = {'response_body': response.body};
    }

    return AppError(
      code: errorCode,
      message: errorMessage,
      details: errorDetails,
    );
  }
}

/// Extended API Client specifically for Extended Exchange
class ExtendedApiClient extends BaseApiClient {
  final String apiKey;

  ExtendedApiClient({
    required String baseUrl,
    required this.apiKey,
    Duration timeout = const Duration(seconds: 30),
  }) : super(
          baseUrl: baseUrl,
          defaultHeaders: {
            'Content-Type': 'application/json',
            'X-API-Key': apiKey,
          },
          timeout: timeout,
        );

  /// Add Extended Exchange specific methods here
  Future<Map<String, dynamic>> getMarketData(String symbol) async {
    return await get<Map<String, dynamic>>(
      '/market/data',
      queryParameters: {'symbol': symbol},
    );
  }

  Future<Map<String, dynamic>> placeOrder(Map<String, dynamic> orderData) async {
    return await post<Map<String, dynamic>>(
      '/orders',
      body: orderData,
    );
  }

  Future<List<Map<String, dynamic>>> getOrderHistory() async {
    final response = await get<Map<String, dynamic>>('/orders/history');
    return (response['orders'] as List).cast<Map<String, dynamic>>();
  }

  Future<Map<String, dynamic>> getBalance() async {
    return await get<Map<String, dynamic>>('/account/balance');
  }
}